File: Crypto/dpdp.py
import random
from hashlib import sha256
from typing import List, Dict, Tuple

from py_ecc.optimized_bn128 import (
    FQ,
    G1, G2,
    pairing,
    add,
    multiply,
    is_inf,
    Z1
)

from common.datastructures import DPDPParams, DPDPTags, DPDPProof
from utils import sha256_hex
from merkle import MerkleTree
from crypto import CURVE_ORDER as curve_order  # 使用 BN128 曲线阶

G1_IDENTITY = Z1

# 对 BN128：使用简化的 hash-to-G1（哈希为标量后乘 G1）
def hash_to_G1(message: bytes):
    k = int.from_bytes(sha256(message).digest(), "big") % curve_order
    return multiply(G1, k)

def chunk_to_int(chunk: bytes) -> int:
    """
    将文件块转换为模曲线阶的整数。
    """
    h = sha256_hex(chunk)
    return int(h, 16)

class dPDP:
    """基于 BLS 签名的 dPDP 方案实现."""

    @staticmethod
    def key_gen(security: int = 256) -> DPDPParams:
        """生成 dPDP 参数和密钥."""
        sk_alpha = random.randint(1, curve_order - 1)
        g = G2
        u = G1
        pk_beta = multiply(g, sk_alpha)
        return DPDPParams(
            g=g,
            u=u,
            pk_beta=pk_beta,
            sk_alpha=sk_alpha
        )

    @staticmethod
    def tag_file(params: DPDPParams, file_chunks: List[bytes]) -> DPDPTags:
        """为每个文件块生成 dPDP 标签."""
        tags: List[Tuple[FQ, FQ, FQ]] = []
        for i, chunk in enumerate(file_chunks):
            b_i = chunk_to_int(chunk)
            b_i %= curve_order
            h_i = hash_to_G1(str(i).encode())
            term1 = h_i
            term2 = multiply(params.u, int(b_i))
            base_point = add(term1, term2)
            sigma_i = multiply(base_point, int(params.sk_alpha))
            tags.append(sigma_i)
        return DPDPTags(tags=tags)
    
    @staticmethod
    def gen_chal(prev_hash: str, timestamp: int, tags: DPDPTags, m: int | None = None) -> List[Tuple[int, int]]:
        """
        依据(prev_hash, timestamp)确定性生成公开挑战集合。
        返回[(i, v_i)]，其中 i∈[0,n)，v_i∈Z_r。
        """
        n = len(tags.tags)
        if n == 0:
            return []
        count = m if m is not None else max(1, (timestamp % n) or 5)
        chal: List[Tuple[int, int]] = []
        for j in range(count):
            seed = f"{prev_hash}:{timestamp}:{j}".encode()
            i = int(sha256(seed).hexdigest(), 16) % n
            v_i = int(sha256(b"chal|" + seed).hexdigest(), 16) % curve_order
            chal.append((i, v_i))
        return chal

    @staticmethod
    def gen_contributions(tags: DPDPTags, file_chunks: Dict[int, bytes], challenge: List[Tuple[int, int]]) -> List[Tuple[int, int, Tuple[FQ, FQ, FQ]]]:
        """
        生成未聚合的分片贡献 (i, mu_i, sigma_i) 列表，其中
        mu_i = v_i * b_i (mod r), sigma_i = v_i * tag_i。
        """
        contributions: List[Tuple[int, int, Tuple[FQ, FQ, FQ]]] = []
        for i, v_i in challenge:
            if i >= len(tags.tags) or i not in file_chunks:
                raise ValueError(f"索引 {i} 在提供的块/标签中未找到，无法生成贡献")
            b_i = chunk_to_int(file_chunks[i]) % curve_order
            sigma_i = tuple(int(x) for x in tags.tags[i])
            sigma_i = tuple(FQ(x) for x in sigma_i)
            mu_i = (v_i * b_i) % curve_order
            sigma_i_scaled = multiply(sigma_i, v_i)
            contributions.append((i, mu_i, sigma_i_scaled))
        return contributions

    @staticmethod
    def gen_proof(tags: DPDPTags, file_chunks: Dict[int, bytes], challenge: List[Tuple[int, int]]) -> DPDPProof:
        """为挑战生成 dPDP 聚合证明。"""
        agg_mu = 0
        agg_sigma = G1_IDENTITY
        for i, v_i in challenge:
            if i >= len(tags.tags):
                raise ValueError(f"索引 {i} 在提供的块/标签中未找到，无法生成证明")
            b_i = chunk_to_int(file_chunks[i])
            sigma_i = tuple(int(x) for x in tags.tags[i])
            sigma_i = tuple(FQ(x) for x in sigma_i)
            agg_mu = (agg_mu + (v_i * (b_i % curve_order))) % curve_order
            agg_sigma = add(agg_sigma, multiply(sigma_i, int(v_i)))
        return DPDPProof(mu=agg_mu, sigma=agg_sigma)

    @staticmethod
    def check_proof(params: DPDPParams, proof: DPDPProof, challenge: List[Tuple[int, int]]) -> bool:
        """验证 dPDP 证明."""
        sigma = tuple(int(x) for x in proof.sigma)
        sigma = tuple(FQ(x) for x in sigma)
        if is_inf(sigma):
            return not challenge
        lhs = pairing(params.g, sigma)
        agg_h = G1_IDENTITY
        for i, v_i in challenge:
            h_i = hash_to_G1(str(i).encode())
            agg_h = add(agg_h, multiply(h_i, int(v_i)))
        u = tuple(int(x) for x in params.u)
        u = tuple(FQ(x) for x in u)
        mu_u = multiply(u, int(proof.mu))
        rhs_g1_point = add(agg_h, mu_u)
        rhs = pairing(params.pk_beta, rhs_g1_point)
        return lhs == rhs

    @staticmethod
    def VerifyProofWithMerkle(
        params: DPDPParams,
        proof: DPDPProof,
        challenge: List[Tuple[int, int]],
        challenged_data: Dict[int, Tuple[bytes, List[Tuple[str, str]]]],
        merkle_root: str
    ) -> bool:
        """
        验证 dPDP 证明及被挑战数据块的 Merkle 证明。

        此函数供领导者/验证者使用，执行完整验证流程：
        1. 验证每个被挑战数据块的 Merkle 证明。
        2. 从已验证的数据块中重新计算聚合哈希 ('mu')。
        3. 检查重新计算的 'mu' 是否与 dPDP 证明中的 'mu' 匹配。
        4. 执行最终的 dPDP 密码学检查。
        """
        # 步骤 1 & 2: 验证 Merkle 证明并重新计算 mu
        recomputed_mu = 0
        
        # 确保证明数据与挑战一一对应
        if {i for i, v in challenge} != set(challenged_data.keys()):
            return False

        for i, v_i in challenge:
            chunk, merkle_proof = challenged_data[i]

            # 验证数据块的 Merkle 证明
            leaf = sha256_hex(chunk)
            if not MerkleTree.verify(leaf, i, merkle_proof, merkle_root):
                return False  # Merkle 证明无效

            # Merkle 证明有效, 使用数据块更新 mu
            b_i = chunk_to_int(chunk)
            recomputed_mu = (recomputed_mu + v_i * b_i) % curve_order

        # 步骤 3: 检查重新计算的 mu 是否与证明中的 mu 匹配
        if recomputed_mu != proof.mu:
            return False  # 聚合数据块哈希 ('mu') 无效

        # 步骤 4: 执行最终的密码学检查
        return dPDP.check_proof(params, proof, challenge)

File: src/crypto/dpdp.rs
use std::collections::HashMap;

use ark_bn254::{Fr, G1Affine, G1Projective, G2Affine, G2Projective};
use ark_ec::{AffineRepr, PrimeGroup};
use ark_ff::{PrimeField, Zero};
use num_bigint::{BigUint, ToBigUint};
use num_traits::cast::ToPrimitive;
use rand::RngCore;

use crate::common::datastructures::{DPDPParams, DPDPProof, DPDPTags};
use crate::crypto::folding::{dpdp_verification_relaxed_r1cs, RelaxedR1CS};
use crate::crypto::{curve_order, deserialize_g1, serialize_g1};
use crate::merkle::MerkleTree;
use crate::utils::{hash_to_field, sha256_hex};

/// Merkle 证明路径，按从叶子到根的顺序存储相邻节点及其方向。
pub type MerkleProofPath = Vec<(String, char)>;

/// 挑战索引到数据块与 Merkle 证明的映射。
pub type ChallengedChunkData = HashMap<usize, (Vec<u8>, MerkleProofPath)>;

/// 将任意消息哈希到 BN254 G1 群，作为 dPDP 中的基点。
pub fn hash_to_g1(message: &[u8]) -> G1Projective {
    let field_elem = hash_to_field(message);
    let mut bytes = field_elem.to_bytes_be();
    if bytes.len() > 32 {
        bytes = bytes[bytes.len() - 32..].to_vec();
    }
    if bytes.len() < 32 {
        let mut padded = vec![0u8; 32 - bytes.len()];
        padded.extend_from_slice(&bytes);
        bytes = padded;
    }
    let scalar = Fr::from_be_bytes_mod_order(&bytes);
    let mut point: G1Projective = G1Affine::generator().into();
    point = point.mul_bigint(scalar.into_bigint());
    if point.is_zero() {
        point = G1Affine::generator().into();
    }
    point
}

/// 生成非零随机标量，用于模拟私钥或挑战系数。
fn random_scalar() -> BigUint {
    let order = curve_order();
    let mut rng = rand::thread_rng();
    loop {
        let mut buf = [0u8; 32];
        rng.fill_bytes(&mut buf);
        let candidate = BigUint::from_bytes_be(&buf) % &order;
        if !candidate.is_zero() {
            return candidate;
        }
    }
}

/// 将 `BigUint` 裁剪到曲线阶范围内并转换为 `Fr` 元素。
fn biguint_to_fr(value: &BigUint) -> Fr {
    let mut bytes = value.to_bytes_be();
    if bytes.len() > 32 {
        bytes = bytes[bytes.len() - 32..].to_vec();
    }
    if bytes.len() < 32 {
        let mut padded = vec![0u8; 32 - bytes.len()];
        padded.extend_from_slice(&bytes);
        bytes = padded;
    }
    Fr::from_be_bytes_mod_order(&bytes)
}

/// 将原始数据块映射到大整数，供 dPDP 计算使用。
fn chunk_to_int(chunk: &[u8]) -> BigUint {
    let h = sha256_hex(chunk);
    BigUint::parse_bytes(h.as_bytes(), 16).unwrap_or_default()
}

/// dPDP 协议的高阶接口封装。
pub struct DPDP;

#[derive(Debug, Clone)]
pub struct DPDPVerificationOutput {
    /// 证明是否通过验证。
    pub valid: bool,
    /// 供 Nova 折叠器继续递归的松弛 R1CS 描述。
    pub circuit: RelaxedR1CS<Fr>,
}

impl DPDP {
    /// 生成 dPDP 公私钥参数。
    pub fn key_gen() -> DPDPParams {
        let sk_alpha = random_scalar();
        let g: G2Projective = G2Affine::generator().into();
        let u: G1Projective = G1Affine::generator().into();
        let sk_fr = biguint_to_fr(&sk_alpha);
        let pk_beta = g.mul_bigint(sk_fr.into_bigint());
        DPDPParams {
            g,
            u,
            pk_beta,
            sk_alpha,
        }
    }

    /// 使用 dPDP 私钥为文件块生成标签。
    pub fn tag_file(params: &DPDPParams, file_chunks: &[Vec<u8>]) -> DPDPTags {
        let mut tags_bytes = Vec::with_capacity(file_chunks.len());
        let sk_fr = biguint_to_fr(&params.sk_alpha);
        for (i, chunk) in file_chunks.iter().enumerate() {
            let mut b_i = chunk_to_int(chunk);
            let order = curve_order();
            b_i %= &order;
            let h_i = hash_to_g1(i.to_string().as_bytes());
            let term2 = params.u.mul_bigint(biguint_to_fr(&b_i).into_bigint());
            let base = h_i + term2;
            let sigma = base.mul_bigint(sk_fr.into_bigint());
            tags_bytes.push(serialize_g1(&sigma));
        }
        DPDPTags { tags: tags_bytes }
    }

    /// 根据链上状态生成挑战向量。
    pub fn gen_chal(
        prev_hash: &str,
        timestamp: u64,
        tags: &DPDPTags,
        m: Option<usize>,
    ) -> Vec<(usize, BigUint)> {
        if tags.is_empty() {
            return Vec::new();
        }
        let n = tags.tags.len();
        let count = if let Some(m) = m {
            m
        } else {
            let mut base = (timestamp as usize) % n;
            if base == 0 {
                base = 5;
            }
            base = base.max(1);
            base
        };
        let order = curve_order();
        let mut challenges = Vec::new();
        for j in 0..count {
            let seed = format!("{}:{}:{}", prev_hash, timestamp, j);
            let idx = (hash_to_field(seed.as_bytes()) % n.to_biguint().unwrap())
                .to_usize()
                .unwrap_or(0)
                % n;
            let value_seed = format!("chal|{}", seed);
            let v_i = hash_to_field(value_seed.as_bytes()) % &order;
            challenges.push((idx, v_i));
        }
        challenges
    }

    /// 针对挑战向量计算每个数据块的贡献值。
    pub fn gen_contributions(
        tags: &DPDPTags,
        file_chunks: &HashMap<usize, Vec<u8>>,
        challenge: &[(usize, BigUint)],
    ) -> Vec<(usize, BigUint, Vec<u8>)> {
        let order = curve_order();
        let mut contributions = Vec::new();
        for (i, v_i) in challenge {
            let chunk = file_chunks
                .get(i)
                .expect("missing chunk for challenge index");
            let mut b_i = chunk_to_int(chunk);
            b_i %= &order;
            let tag_bytes = tags.tags.get(*i).expect("missing tag for challenge index");
            let sigma_i = deserialize_g1(tag_bytes);
            let mu_i = (v_i * &b_i) % &order;
            let sigma_scaled = sigma_i.mul_bigint(biguint_to_fr(v_i).into_bigint());
            contributions.push((*i, mu_i, serialize_g1(&sigma_scaled)));
        }
        contributions
    }

    /// 聚合挑战，生成完整的 dPDP 证明。
    pub fn gen_proof(
        tags: &DPDPTags,
        file_chunks: &HashMap<usize, Vec<u8>>,
        challenge: &[(usize, BigUint)],
    ) -> DPDPProof {
        let order = curve_order();
        let mut agg_mu = BigUint::from(0u32);
        let mut agg_sigma = G1Projective::zero();
        for (i, v_i) in challenge {
            let chunk = file_chunks.get(i).expect("missing chunk");
            let mut b_i = chunk_to_int(chunk);
            b_i %= &order;
            let tag_bytes = tags.tags.get(*i).expect("missing tag");
            let sigma_i = deserialize_g1(tag_bytes);
            agg_mu = (agg_mu + (v_i * &b_i)) % &order;
            agg_sigma += sigma_i.mul_bigint(biguint_to_fr(v_i).into_bigint());
        }
        DPDPProof {
            mu: agg_mu.to_string(),
            sigma: serialize_g1(&agg_sigma),
        }
    }

    /// 仅检查证明是否有效。
    pub fn check_proof(
        params: &DPDPParams,
        proof: &DPDPProof,
        challenge: &[(usize, BigUint)],
    ) -> bool {
        Self::check_proof_with_relaxed(params, proof, challenge).valid
    }

    /// 返回验证结果及对应的松弛 R1CS 实例。
    pub fn check_proof_with_relaxed(
        params: &DPDPParams,
        proof: &DPDPProof,
        challenge: &[(usize, BigUint)],
    ) -> DPDPVerificationOutput {
        let sigma = deserialize_g1(&proof.sigma);
        if sigma.is_zero() {
            let (circuit, _) = dpdp_verification_relaxed_r1cs(params, proof, challenge);
            return DPDPVerificationOutput {
                valid: challenge.is_empty(),
                circuit,
            };
        }

        let (circuit, valid) = dpdp_verification_relaxed_r1cs(params, proof, challenge);
        DPDPVerificationOutput { valid, circuit }
    }

    /// 验证带有 Merkle 证明的 dPDP 响应。
    pub fn verify_with_merkle(
        params: &DPDPParams,
        proof: &DPDPProof,
        challenge: &[(usize, BigUint)],
        challenged_data: &ChallengedChunkData,
        merkle_root: &str,
    ) -> bool {
        let order = curve_order();
        let mut recomputed_mu = BigUint::from(0u32);
        if challenge.len() != challenged_data.len() {
            return false;
        }
        for (i, v_i) in challenge {
            let Some((chunk, proof_path)) = challenged_data.get(i) else {
                return false;
            };
            let leaf = sha256_hex(chunk);
            if !MerkleTree::verify(&leaf, *i, proof_path, merkle_root) {
                return false;
            }
            let mut b_i = chunk_to_int(chunk);
            b_i %= &order;
            recomputed_mu = (recomputed_mu + v_i * &b_i) % &order;
        }
        let mu_big = BigUint::parse_bytes(proof.mu.as_bytes(), 10).unwrap_or_default();
        if mu_big != recomputed_mu {
            return false;
        }
        Self::check_proof(params, proof, challenge)
    }
}
